server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/RequestForwardingUtil.kt

```diff
@@ -0,0 +1,97 @@
+package com.xebia.functional.xef.server.http.routes
+
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.http.content.*
+import io.ktor.server.application.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.util.*
+import  io.ktor.http.URLBuilder
+import io.ktor.util.pipeline.*
+import io.ktor.utils.io.jvm.javaio.*
+
+private const val OAI_URL = "https://api.openai.com"
+
+/**
+ * Retrieves the path from the incoming request and
+ * combines it with the provider specific host url.
+ * As we are copying the API from the providers,
+ * everything including the path structure has follow the API contract.
+ */
+private fun buildProviderUrlFromRequest(request: ApplicationRequest) =
+    URLBuilder().takeFrom(OAI_URL).apply {
+        encodedPath = request.path()
+    }.build()
+
+/**
+ * Makes a request to the provider forwarding headers and request body
+ * from the incoming request.
+ * The provider's response is then forwarded as a response of the server.
+ *
+ * Takes in [body] as Bytes and responds in Bytes.
+ * No messing around with char sets. Just forwarding raw bytes.
+ *
+ * @return provider's response
+ */
+internal suspend fun PipelineContext<Unit, ApplicationCall>.forwardToProvider(
+    client: HttpClient,
+    stream: Boolean = false,
+    url: String = buildProviderUrlFromRequest(call.request).toString(),
+): HttpResponse = if (stream) {
+    client
+        .prepareRequest { prepareRequest(call, url) }
+        .execute { res -> call.forwardResponse(res, stream = stream); res }
+} else {
+    val response = client.request { prepareRequest(call, url) }
+    call.forwardResponse(response, stream = stream)
+    response
+}
+
+private suspend fun HttpRequestBuilder.prepareRequest(
+    call: ApplicationCall,
+    urlParam: String,
+) {
+    url(urlParam)
+    method = call.request.httpMethod
+    headers.copyFrom(call.request.headers) // copy headers
+    url.parameters.appendAll(call.request.queryParameters) // copy parameters
+
+    val body = call
+        .receiveChannel()
+        .toByteArray()
+        .let { ByteArrayContent(it, contentType = null) }
+    setBody(body)
+}
+
+private suspend fun ApplicationCall.forwardResponse(
+    providerResponse: HttpResponse,
+    stream: Boolean,
+) {
+    response.headers.copyFrom(providerResponse.headers)
+
+    if(stream) {
+        respondOutputStream {
+            providerResponse
+                .bodyAsChannel()
+                .copyTo(this@respondOutputStream)
+        }
+    } else {
+        respond(providerResponse.status, providerResponse.readBytes())
```

Comment:
@Intex32, is this reading the body twice (https://github.com/xebia-functional/xef/pull/482/files#diff-5767cc2ab5cd1d7765af84cd0859fec6baf453ecef7e81b518f58e0890a4b692R43)? The same for the streaming

Reply:
Yes, first your reading and parsing it to a JsonObject (for monitoring or so) and in the `forward` function the input is received a second time as raw bytes and passed forward. I found this to be the cleanest solution, as both is independent.